### PR TITLE
Mocking hacks

### PR DESCRIPTION
**Pull request by ArcturusZhang**
This PR is just to brainstorm options as is not intended to be reviewed / merged

Related issue: https://github.com/Azure/autorest.csharp/issues/3493

This PR is trying to discover some convenient layer for our customers to get the mocking of our extension methods.

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: eng/Packages.Data.props [hunk 1]
```diff
@@ -102,6 +102,8 @@
     <PackageReference Update="Azure.Storage.Queues" Version="12.14.0" />
     <PackageReference Update="Azure.ResourceManager" Version="1.5.0" />
 
+    <PackageReference Update="Moq" Version="4.18.2" />
```
Comment (jsquire) [inline eng/Packages.Data.props:105]: Moq already appears under the Test project section (L258).   We do not want to add it here, as it should not be referenced by our non-test projects.

File: sdk/resources/Azure.ResourceManager.Resources/src/Generated/Extensions/MockingExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+using System;
+using Moq;
+using Mock = Moq.Mock;
+using System.Reflection;
+
+namespace Azure.ResourceManager.Resources.Mocking
+{
+    internal static class MockingExtensions
+    {
+        internal static void SetupAzureExtensionMethod<T>(this Mock<T> mock, Expression<Action<T>> expression) where T : ArmResource
```
Comment (jsquire) [inline sdk/resources/Azure.ResourceManager.Resources/src/Generated/Extensions/MockingExtensions.cs:15]: What if a customer is using a different mocking library?   For example, `NSubstitute`, `FakeItEasy`, `RhinoMocks`, `JustMock`

We do not want to offer built-in support for a specific mocking library and remove customer choice or provide a degraded experience for those who prefer other libraries.

File: eng/Directory.Build.Common.targets [hunk 3]
```diff
@@ -214,6 +214,7 @@
 
     <PackageReference Include="Azure.Core" />
     <PackageReference Include="System.Text.Json" />
+    <PackageReference Include="Moq" />
```
Comment (jsquire) [inline eng/Directory.Build.Common.targets:217]: We should not be including `Moq` in a non-test project.

**Review by jsquire — CHANGES_REQUESTED** [review]

**Comment by jsquire** [comment]
I don't know if this is just experimentation or the intention was to actually move forward with this as part of the Azure SDK.   Please add a meaningful description to all pull requests that you open.  Anyone should be able to understand your intent and the scope of your work by reading the PR.

**Comment by ArcturusZhang** [comment]
> I don't know if this is just experimentation or the intention was to actually move forward with this as part of the Azure SDK. Please add a meaningful description to all pull requests that you open. Anyone should be able to understand your intent and the scope of your work by reading the PR.

Yes this is only an experiment - when it is ready to review I will add more descriptions

**Comment by jsquire** [comment]
> Yes this is only an experiment - when it is ready to review I will add more descriptions

Please add a meaningful description for every PR you open, when you open it.   The intent should be clear upfront.  Let's not wait until things come out of draft status.   Anyone looking at PRs in our repository should be able to open any pull request - draft or not - and understand why it exists and what it is attempting to do.

File: sdk/resources/Azure.ResourceManager.Resources/tests/MockingTests.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+            var asyncResult = await tenant.CalculateDeploymentTemplateHashAsync(mockTemplate, default);
+            Assert.IsNull(asyncResult);
+            //mock.Setup(sub => sub.GetCachedClient(It.IsAny<Func<ArmClient, TenantResourceExtensionClient>>()));//.Returns(armDeploymentResourceExtensionMock.Object);
+
+            //Assert.Throws<Exception>(() => tenant.GetArmDeployment("", default));
+        }
+
+        [Test]
+        public async Task TestUsingAzureMock()
```
Comment (ArcturusZhang) [inline sdk/resources/Azure.ResourceManager.Resources/tests/MockingTests.cs:42]: This is the instance way.
Customer needs to use `AzureMock<T>` instead when mocking azure APIs.
This AzureMock not only works for extension methods, it also works just like Mock for normal instance method.

**Comment by ArcturusZhang** [comment]
The things we need to override to redirect: 
- `public ISetup<T> Setup(Expression<Action<T>> expression)` in Mock<T>
- `public ISetup<T, TResult> Setup<TResult>(Expression<Func<T, TResult>> expression)` in Mock<T>
The adaptors we need:
- `ISetup<TMock, TResult>` (60 methods)
- `IReturnsResult<T>` (40 methods)

**Pull request closed without merging**